### ISSUE TITLE
`unsafeRun` failure handling when E <: Throwable

### ISSUE THREAD

**Issue by dkarlinsky** [zio/zio#2546]
The implementation of `Runtime.unsafeRun()` wraps any cause in `FiberFailure`.
I think this not a very intuitive behavior in case of `E <: Throwable`, especially for beginners that will naturally reach for this method, as it seems the most straight forward.
That's what we did everywhere in our projects, and it took us some time to figure our why we were always getting `FiberFailure` on every error. Since we mostly track our production exceptions by class, this wasn't ideal.

What I came up instead is something like this:

```scala
def unsafeRunWithThrow[A, E <: Throwable](effect: ZIO[R, E, A]): A = {
      runtime.unsafeRunSync(effect).fold( cause =>
         throw cause.failureOption.getOrElse(FiberFailure(cause)),
        identity
      )
}
```

This does what I initially expected `unsafeRunSync` to do - throw the exception that caused the failure.

We do loose the trace for the `failure` causes, which is unfortunate. 
I did come up for a solution for this too, which is a bit hacky but works well for us.
I create a special Throwable with `cause.prettyPrint` as the message, and make it the root `Throwable.cause` for the original exception in the failure cause (and actually `die` causes as well).
See code bellow.

What I would like to do is to fold this behavior into `unsafeRun`. We could use a `Manifest[E]` and handle `E <: Throwable` differently. 
Or if not, we could add a new method with this behavior to Runtime.

WDYT?

-------------
So my `unsafeRunWithThrow` actually looks like this:
```scala
def unsafeRunWithThrow[A, E <: Throwable](effect: ZIO[R, E, A]): A = {
      runtime.unsafeRunSync(effect).fold( cause =>
         throw cause.failureOption.map(attachTrace(cause))
            .orElse(cause.dieOption.map(attachTrace(cause)))
            .getOrElse(FiberFailure(cause)),
        identity
      )
}

private def attachTrace[E <: Throwable](cause: Cause[E])(e: E): E = {
      if(cause.traces.nonEmpty) {
        ignoring(classOf[Throwable]) {
          val rootCause = ExceptionUtils.getRootCause(e)
          forceCause(FiberTraceThrowable(cause.prettyPrint), rootCause)
        }
      }
      e
}

private def forceCause(fiberFailure: Throwable,
                         rootCause: Throwable) = {
    val causeField = classOf[Throwable].getDeclaredField("cause")
    causeField.setAccessible(true)
    causeField.set(rootCause, fiberFailure)
  }

private case class FiberTraceThrowable(trace: String) extends Throwable(null, null, true, false) {
    override final def getMessage: String = trace
}
```

**Comment by dkarlinsky** [comment]
@jdegoes WDYT?

**Comment by jdegoes** [comment]
I think this would make sense in cats-interop.

What could make sense in ZIO is `unsafeRunTask` and variants that require the `E` to be `Throwable`, and then `throw` it.

**Comment by jdegoes** [comment]
Ah, I see you were proposing something similar. The error need only be `Throwable`, you don't need `R <: Throwable` due to covariance.

**Comment by dkarlinsky** [comment]
Thanks @jdegoes, `unsafeRunTask` makes sense.
What about attaching fiber dump info to thrown exceptions as a cause?

**Comment by jdegoes** [comment]
@dkarlinsky Yes, I'd attach the `FiberFailure` as a cause.

**Comment by neko-kai** [comment]
Would forcing the `cause` reflectively like that work on Scala.js & Native? Also, you'd have to put the original `cause` of the error somewhere, since replacing it would modify both the outer throwable, and the same throwable inside the Cause tree

**Comment by dkarlinsky** [comment]
@neko-kai Good point about Scala.js and Native. I think worst case, we just skip this part for these platforms.
As to the `cause`, I'm adding a new throwable at the end of the cause chain, not replacing the immediate cause of the throwable. It does modify the throwable and thus the original `Cause[E]`, since there is no way (I know of) to make a modified copy of an unknown throwable...
This is why I'm not using `FiberFailure` for the appended `cause` - this would create a cycle.
Instead I used a simple stackless throwable with a the pretty printed `Cause[E]` as the message.

**Comment by iravid** [comment]
Fixed in #2600 - thanks @dkarlinsky!
